assign a func var from binary method

The following program `sample.go` triggers a panic:

```go
package main

import "net"

var lookupHost = net.DefaultResolver.LookupHost

func main() {
    println(lookupHost != nil)
}
```

Expected result:
```console
$ go run ./sample.go
true
```

Got:
```console
$ yaegi ./sample.go
./sample.go:7:18: panic
panic: reflect.Set: value of type func(context.Context, string) ([]string, error) is not assignable to type func(*net.Resolver, context.Context, string) ([]string, error) [recovered]
        panic: reflect.Set: value of type func(context.Context, string) ([]string, error) is not assignable to type func(*net.Resolver, context.Context, string) ([]string, error)

goroutine 1 [running]:
github.com/containous/yaegi/interp.runCfg.func1(0xc000256000, 0xc000170b00)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:114 +0x218
panic(0xc9da60, 0xc00015e5b0)
        /usr/lib/go/src/runtime/panic.go:679 +0x1b2
reflect.Value.assignTo(0xdbf8a0, 0x15269d8, 0xa93, 0xddc278, 0xb, 0xc000164280, 0x0, 0x2b, 0x0, 0x0)
        /usr/lib/go/src/reflect/value.go:2403 +0x432
reflect.Value.Set(0xc000164280, 0xc000160120, 0x193, 0xdbf8a0, 0x15269d8, 0xa93)
        /usr/lib/go/src/reflect/value.go:1537 +0xbd
github.com/containous/yaegi/interp.assign.func3(0xc000256000, 0xc0002543c0)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:304 +0x14f
github.com/containous/yaegi/interp.runCfg(0xc000170b00, 0xc000256000)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:120 +0x7a
github.com/containous/yaegi/interp.(*Interpreter).run(0xc00024a6e0, 0xc000170200, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:98 +0x239
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc00024a6e0, 0xc00025e000, 0x7e, 0x27e, 0xc00025e000, 0x7e, 0x0, 0x1)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:364 +0x3ef
main.main()
        /home/marc/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:82 +0x42e
```

--- a/_test/method29.go
+++ b/_test/method29.go
@@ -2,15 +2,15 @@ package main
 
 import (
 	"context"
-	"fmt"
 	"net"
 )
 
 var lookupHost = net.DefaultResolver.LookupHost
 
 func main() {
-	fmt.Println(lookupHost(context.Background(), "localhost"))
+	res, err := lookupHost(context.Background(), "localhost")
+	println(len(res) > 0, err == nil)
 }
 
 // Output:
-// [::1 127.0.0.1] <nil>
+// true true

--- a/_test/method29.go
+++ b/_test/method29.go
@@ -1,0 +1,16 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+)
+
+var lookupHost = net.DefaultResolver.LookupHost
+
+func main() {
+	fmt.Println(lookupHost(context.Background(), "localhost"))
+}
+
+// Output:
+// [::1 127.0.0.1] <nil>

--- a/_test/method30.go
+++ b/_test/method30.go
@@ -1,0 +1,20 @@
+package main
+
+type T struct {
+	Name string
+}
+
+func (t *T) foo(a string) string {
+	return t.Name + a
+}
+
+var g = &T{"global"}
+
+var f = g.foo
+
+func main() {
+	println(f("-x"))
+}
+
+// Output:
+// global-x

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -410,7 +410,11 @@ func (interp *Interpreter) cfg(root *node) ([]*node, error) {
 						if src.typ, err = nodeType(interp, sc, src); err != nil {
 							return
 						}
-						dest.typ = src.typ
+						if src.typ.isBinMethod {
+							dest.typ = &itype{cat: valueT, rtype: src.typ.methodCallType()}
+						} else {
+							dest.typ = src.typ
+						}
 					}
 					if dest.typ.sizedef {
 						dest.typ.size = arrayTypeLen(src)
@@ -1097,7 +1101,7 @@ func (interp *Interpreter) cfg(root *node) ([]*node, error) {
 					n.val = method.Index
 					n.gen = getIndexBinMethod
 					n.recv = &receiver{node: n.child[0]}
-					n.typ = &itype{cat: valueT, rtype: method.Type}
+					n.typ = &itype{cat: valueT, rtype: method.Type, isBinMethod: true}
 				case n.typ.rtype.Kind() == reflect.Ptr:
 					if field, ok := n.typ.rtype.Elem().FieldByName(n.child[1].ident); ok {
 						n.typ = &itype{cat: valueT, rtype: field.Type}

--- a/interp/gta.go
+++ b/interp/gta.go
@@ -60,6 +60,9 @@ func (interp *Interpreter) gta(root *node, rpath string) ([]*node, error) {
 					err = n.cfgErrorf("use of untyped nil")
 					return false
 				}
+				if typ.isBinMethod {
+					typ = &itype{cat: valueT, rtype: typ.methodCallType(), isBinMethod: true}
+				}
 				sc.sym[dest.ident] = &symbol{kind: varSym, global: true, index: sc.add(typ), typ: typ, rval: val}
 				if n.anc.kind == constDecl {
 					sc.sym[dest.ident].kind = constSym

--- a/interp/run.go
+++ b/interp/run.go
@@ -828,7 +828,9 @@ func callBin(n *node) {
 	// method signature obtained from reflect.Type include receiver as 1st arg, except for interface types
 	rcvrOffset := 0
 	if recv := n.child[0].recv; recv != nil && recv.node.typ.TypeOf().Kind() != reflect.Interface {
-		rcvrOffset = 1
+		if funcType.NumIn() > len(child) {
+			rcvrOffset = 1
+		}
 	}
 
 	for i, c := range child {
@@ -867,7 +869,6 @@ func callBin(n *node) {
 			case interfaceT:
 				values = append(values, genValueInterfaceValue(c))
 			default:
-				//values = append(values, genValue(c))
 				values = append(values, genInterfaceWrapper(c, defType))
 			}
 		}

--- a/interp/type.go
+++ b/interp/type.go
@@ -103,22 +103,23 @@ type structField struct {
 
 // itype defines the internal representation of types in the interpreter
 type itype struct {
-	cat        tcat          // Type category
-	field      []structField // Array of struct fields if structT or interfaceT
-	key        *itype        // Type of key element if MapT or nil
-	val        *itype        // Type of value element if chanT, mapT, ptrT, aliasT, arrayT or variadicT
-	arg        []*itype      // Argument types if funcT or nil
-	ret        []*itype      // Return types if funcT or nil
-	method     []*node       // Associated methods or nil
-	name       string        // name of type within its package for a defined type
-	path       string        // for a defined type, the package import path
-	size       int           // Size of array if ArrayT
-	rtype      reflect.Type  // Reflection type if ValueT, or nil
-	incomplete bool          // true if type must be parsed again (out of order declarations)
-	untyped    bool          // true for a literal value (string or number)
-	sizedef    bool          // true if array size is computed from type definition
-	node       *node         // root AST node of type definition
-	scope      *scope        // type declaration scope (in case of re-parse incomplete type)
+	cat         tcat          // Type category
+	field       []structField // Array of struct fields if structT or interfaceT
+	key         *itype        // Type of key element if MapT or nil
+	val         *itype        // Type of value element if chanT, mapT, ptrT, aliasT, arrayT or variadicT
+	arg         []*itype      // Argument types if funcT or nil
+	ret         []*itype      // Return types if funcT or nil
+	method      []*node       // Associated methods or nil
+	name        string        // name of type within its package for a defined type
+	path        string        // for a defined type, the package import path
+	size        int           // Size of array if ArrayT
+	rtype       reflect.Type  // Reflection type if ValueT, or nil
+	incomplete  bool          // true if type must be parsed again (out of order declarations)
+	untyped     bool          // true for a literal value (string or number)
+	sizedef     bool          // true if array size is computed from type definition
+	isBinMethod bool          // true if the type refers to a bin method function
+	node        *node         // root AST node of type definition
+	scope       *scope        // type declaration scope (in case of re-parse incomplete type)
 }
 
 // nodeType returns a type definition for the corresponding AST subtree
@@ -482,7 +483,7 @@ func nodeType(interp *Interpreter, sc *scope, n *node) (*itype, error) {
 			if m, _ := lt.lookupMethod(name); m != nil {
 				t, err = nodeType(interp, sc, m.child[2])
 			} else if bm, _, _, ok := lt.lookupBinMethod(name); ok {
-				t = &itype{cat: valueT, rtype: bm.Type}
+				t = &itype{cat: valueT, rtype: bm.Type, isBinMethod: true}
 			} else if ti := lt.lookupField(name); len(ti) > 0 {
 				t = lt.fieldSeq(ti)
 			} else if bs, _, ok := lt.lookupBinField(name); ok {
@@ -807,6 +808,22 @@ func (t *itype) lookupBinField(name string) (s reflect.StructField, index []int,
 		}
 	}
 	return s, index, ok
+}
+
+// methodCallType returns a method function type without the receiver defined.
+// The input type must be a method function type with the receiver as the first input argument.
+func (t *itype) methodCallType() reflect.Type {
+	it := []reflect.Type{}
+	ni := t.rtype.NumIn()
+	for i := 1; i < ni; i++ {
+		it = append(it, t.rtype.In(i))
+	}
+	ot := []reflect.Type{}
+	no := t.rtype.NumOut()
+	for i := 0; i < no; i++ {
+		ot = append(ot, t.rtype.Out(i))
+	}
+	return reflect.FuncOf(it, ot, t.rtype.IsVariadic())
 }
 
 // getMethod returns a pointer to the method definition

